pjlib/src/pj/sock_bsd.c

```diff
@@ -905,4 +905,127 @@ PJ_DEF(pj_status_t) pj_sock_accept( pj_sock_t serverfd,
 }
 #endif	/* PJ_HAS_TCP */
 
+/*
+ * Simulate unix-like socketpair()
+ * Use a pair of IPv4/IPv6 local sockets (listening on 127.0.0.1/::1)
+ */
+static int socketpair_s(int family, int type, int protocol, pj_sock_t sv[2])
+{
+    int status;
+    pj_sock_t lfd = PJ_INVALID_SOCKET;
+    pj_sock_t cfd = PJ_INVALID_SOCKET;
+    pj_str_t loopback;
+    pj_sockaddr sa;
+    int salen;
+
+#if PJ_HAS_TCP
+    PJ_ASSERT_RETURN(type == pj_SOCK_DGRAM() || type == pj_SOCK_STREAM(),
+		     PJ_EINVAL);
+#else
+    PJ_ASSERT_RETURN(type == pj_SOCK_DGRAM(), PJ_EINVAL);
+#endif
+
+    PJ_ASSERT_RETURN(family == pj_AF_UNIX() || family == pj_AF_INET() ||
+			 family == pj_AF_INET6(), PJ_EINVAL);
+
+    if (family == pj_AF_UNIX()) {
+	family = pj_AF_INET();
+    }
+    loopback = family == pj_AF_INET() ? pj_str("127.0.0.1") : pj_str("::1");
+
+    /* listen */
+    status = pj_sock_socket(family, type, protocol, &lfd);
+    if (status != PJ_SUCCESS)
+	goto on_error;
+
+    pj_sockaddr_init(family, &sa, &loopback, 0);
+    salen = pj_sockaddr_get_len(&sa);
+    status = pj_sock_bind(lfd, &sa, salen);
+    if (status != PJ_SUCCESS)
+	goto on_error;
+
+    status = pj_sock_getsockname(lfd, &sa, &salen);
+    if (status != PJ_SUCCESS)
+	goto on_error;
+
+#if PJ_HAS_TCP
+    if (type == pj_SOCK_STREAM()) {
+	status = pj_sock_listen(lfd, 1);
+	if (status != PJ_SUCCESS)
+	    goto on_error;
+    }
+#endif
+
+    /* connect to listen fd */
+    status = pj_sock_socket(family, type, protocol, &cfd);
+    if (status != PJ_SUCCESS)
+	goto on_error;
+    status = pj_sock_connect(cfd, &sa, salen);
+    if (status != PJ_SUCCESS)
+	goto on_error;
+
+    if (type == pj_SOCK_DGRAM()) {
+	status = pj_sock_getsockname(cfd, &sa, &salen);
+	if (status != PJ_SUCCESS)
+	    goto on_error;
+	status = pj_sock_connect(lfd, &sa, salen);
```

Comment:
I think it's simpler to connect cfd to lfd, so we don't have to call getsockname() on cfd. It's also more intuitive (I'm thinking "c" in "cfd" as client).

Reply:
socketpair is two-way connection. so udp both need cfd connect lfd  and lfd connect cfd.